### PR TITLE
limit APNs cluster workers - leasing of the permission to push for APNs AGPUSH-1413

### PR DESCRIPTION
**Pull request by lfryc**
Distributed version of APNs service cache - limiting number of services that can be created across entire cluster.

Tests use 2 clustered instances of WildFly (in Full HA configuration).

https://issues.jboss.org/browse/AGPUSH-1413

### CODE REVIEW

**Comment by lfryc** [comment]
@matzew this builds upon the recent JMS related work

File: README.md [hunk 1]
```diff
@@ -35,6 +35,22 @@ Or for the on-premise version, execute the following steps to get going!
...
+
+In order to test on a cluster of WildFly servers, the default configuration serves pretty well, you just need to change startup script a bit - in following scenario we will use servers colocated on one node with configured port-offset:
+
+    ./bin/standalone.sh -c standalone-full-ha.xml -Djboss.node.name=node1 -Djboss.messaging.cluster.password=somepassword -Djboss.socket.binding.port-offset=100 -Djava.net.preferIPv4Stack=true
+
+And in a second terminal:
+
+    ./bin/standalone.sh -c standalone-full-ha.xml -Djboss.node.name=node2 -Djboss.messaging.cluster.password=somepassword -Djboss.socket.binding.port-offset=200 -Djava.net.preferIPv4Stack=true
+
+Note: on OS X, you may need to enable multicast first:
```
Comment (matzew) [inline README.md:48]: may? :)

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/serviceLease/ApnsServiceHolder.java [hunk 2]
```diff
@@ -35,26 +43,43 @@
...
+    public static final long INSTANCE_ACQUIRING_TIMEOUT = 7500;
+    public static final long DISPOSING_DELAY = 5000;
 
     @Inject
     private ClientInstallationService clientInstallationService;
 
-    public ApnsServiceCache() {
-        super(INSTANCE_LIMIT, INSTANCE_ACQUIRING_TIMEOUT);
+    @Resource(mappedName = "java:/queue/APNsBadgeLeaseQueue")
+    private Queue apnsBadgeLeaseQueue;
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/serviceLease/ApnsServiceHolder.java:63]: badge... is that a good name ? :)

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/serviceLease/ServiceDisposalScheduler.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    private ScheduledExecutorService scheduler;
+    private long terminationTimeout = 10000L;
+
+    /**
+     * Creates a new scheduler on container startup
+     */
+    @PostConstruct
+    public void initialize() {
+        scheduler = Executors.newSingleThreadScheduledExecutor();
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/serviceLease/ServiceDisposalScheduler.java:29]: we could use java ee concurrency ?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/MockServiceCacheForCluster.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package org.jboss.aerogear.unifiedpush.message;
```
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/MockServiceCacheForCluster.java:1]: license header

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/MockServiceCacheForSingleNode.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package org.jboss.aerogear.unifiedpush.message;
```
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/MockServiceCacheForSingleNode.java:1]: header

**Comment by matzew** [comment]
@lfryc can you rebase this off of master, and send a PR against upstream ?

**Comment by lfryc** [comment]
@matzew sure thing, resent to upstream...

**Pull request closed without merging**